File: src/main/java/cn/lnkdoc/sdk/uia/instance/Instance.java
package cn.lnkdoc.sdk.uia.instance;

import cn.lnkdoc.sdk.uia.common.exception.UiaException;
import cn.lnkdoc.sdk.uia.common.property.IUiaProperty;
import cn.lnkdoc.sdk.uia.instance.bjtoon.BjtoonSdkInstance;
import cn.lnkdoc.sdk.uia.instance.bjtoon.property.BjtoonProperty;
import cn.lnkdoc.sdk.uia.instance.jban.JbanSdkInstance;
import cn.lnkdoc.sdk.uia.instance.jban.property.JbanProperty;
import cn.lnkdoc.sdk.uia.instance.wx.WxSdkInstance;
import cn.lnkdoc.sdk.uia.instance.wx.property.WxProperty;
import cn.lnkdoc.sdk.uia.instance.yztoon.YztoonSdkInstance;
import cn.lnkdoc.sdk.uia.instance.yztoon.property.YztoonProperty;
import io.vavr.Function1;

/**
 * @author langkye
 * @since 1.0.0.RELEASE
 */
@SuppressWarnings("ALL")
public enum Instance {
    /**
     * yztoon
     */
    YZTOON(Instance::yztoon, YztoonProperty.class),
    /**
     * bjtoon
     */
    BJTOON(Instance::bjtoon, BjtoonProperty.class),
    /**
     * jban
     */
    JBAN(Instance::jban, JbanProperty.class),
    /**
     * wx
     */
    WX(Instance::wx, WxProperty.class),
    ;

    
    // ~=============================== Instances ===============================~ //

    /**
     * of
     * 
     * @param property property
     * @return instance
     */
    private static YztoonSdkInstance yztoon(IUiaProperty property) {
        return new YztoonSdkInstance((YztoonProperty) property);
    }
    /**
     * of
     *
     * @param property property
     * @return instance
     */
    private static BjtoonSdkInstance bjtoon(IUiaProperty property) {
        return new BjtoonSdkInstance((BjtoonProperty) property);
    }
    /**
     * of
     *
     * @param property property
     * @return instance
     */
    private static JbanSdkInstance jban(IUiaProperty property) {
        return new JbanSdkInstance((JbanProperty) property);
    }
    /**
     * of
     *
     * @param property property
     * @return instance
     */
    private static WxSdkInstance wx(IUiaProperty property) {
        return new WxSdkInstance((WxProperty) property);
    }
    
    
    
    
    // ~=============================== Fileds/Constructor/Getter ===============================~ //
    /**
     * function
     */
    private final Function1<IUiaProperty, ISdkInstance> function;
    /**
     * propertyClazz
     */
    private final Class propertyClazz;


    /**
     * constructor
     * 
     * @param function function
     * @param propertyClazz propertyClazz
     */
    Instance(Function1<IUiaProperty, ISdkInstance> function, Class propertyClazz) {
        this.function = function;
        this.propertyClazz = propertyClazz;
    }

    /**
     * getter
     * 
     * @return function
     */
    public Function1<IUiaProperty, ISdkInstance> getFunction() {
        return function;
    }

    /**
     * getter
     * 
     * @return property clazz
     */
    public Class getPropertyClazz() {
        return propertyClazz;
    }

    // ~=============================== Tools ===============================~ //

    /**
     * of
     * 
     * @param name name
     * @return instance
     */
    public static Instance of(String name) {
        for (Instance instance : Instance.values()) {
            if (instance.toString().equals(name)) {
                return instance;
            }
        }
        throw new UiaException("not match instance for “" + name + "”");
    }

    /**
     * of
     * 
     * @param propertyClazz propertyClazz
     * @return instance
     */
    public static Instance of(Class propertyClazz) {
        for (Instance instance : Instance.values()) {
            if (instance.propertyClazz.equals(propertyClazz)) {
                return instance;
            }
        }
        throw new UiaException("un support property type : " + propertyClazz);
    }
}

File: src/main/kotlin/cn/lnkdoc/sdk/uia/instance/Instance.kt
package cn.lnkdoc.sdk.uia.instance

import cn.lnkdoc.sdk.uia.common.exception.UiaException
import cn.lnkdoc.sdk.uia.common.property.IUiaProperty
import cn.lnkdoc.sdk.uia.instance.alipay.AlipaySdkInstance
import cn.lnkdoc.sdk.uia.instance.alipay.property.AlipayProperty
import cn.lnkdoc.sdk.uia.instance.bjtoon.BjtoonSdkInstance
import cn.lnkdoc.sdk.uia.instance.bjtoon.property.BjtoonProperty
import cn.lnkdoc.sdk.uia.instance.gitee.GiteeSdkInstance
import cn.lnkdoc.sdk.uia.instance.gitee.property.GiteeProperty
import cn.lnkdoc.sdk.uia.instance.github.GithubSdkInstance
import cn.lnkdoc.sdk.uia.instance.github.property.GithubProperty
import cn.lnkdoc.sdk.uia.instance.gitlab.GitlabSdkInstance
import cn.lnkdoc.sdk.uia.instance.gitlab.property.GitlabProperty
import cn.lnkdoc.sdk.uia.instance.google.GoogleSdkInstance
import cn.lnkdoc.sdk.uia.instance.google.property.GoogleProperty
import cn.lnkdoc.sdk.uia.instance.bccastle.BccastleSdkInstance
import cn.lnkdoc.sdk.uia.instance.bccastle.property.BccastleProperty
import cn.lnkdoc.sdk.uia.instance.jban.JbanSdkInstance
import cn.lnkdoc.sdk.uia.instance.jban.property.JbanProperty
import cn.lnkdoc.sdk.uia.instance.weibo.WeiboSdkInstance
import cn.lnkdoc.sdk.uia.instance.weibo.property.WeiboProperty
import cn.lnkdoc.sdk.uia.instance.wx.WxSdkInstance
import cn.lnkdoc.sdk.uia.instance.wx.property.WxProperty
import cn.lnkdoc.sdk.uia.instance.yztoon.YztoonSdkInstance
import cn.lnkdoc.sdk.uia.instance.yztoon.property.YztoonProperty

/**
 * @author langkye
 * @since 1.0.0.RELEASE
 */
@Suppress("unused")
enum class Instance
/**
 * constructor
 *
 * @param function function
 * @param propertyClazz propertyClazz
 */(
    /**
     * function
     */
    val function: (IUiaProperty) -> ISdkInstance,
    /**
     * propertyClazz
     */
    val propertyClazz: Class<*>
) {
    /**
     * yztoon
     */
    YZTOON({ property -> yztoon(property) }, YztoonProperty::class.java),

    /**
     * bjtoon
     */
    BJTOON({ property -> bjtoon(property) }, BjtoonProperty::class.java),

    /**
     * jban
     */
    JBAN({ property -> jban(property) }, JbanProperty::class.java),

    /**
     * wx
     */
    WX({ property -> wx(property) }, WxProperty::class.java),

    /**
     * alipay
     */
    ALIPAY({ property -> alipay(property) }, AlipayProperty::class.java),

    /**
     * github
     */
    GITHUB({ property -> github(property) }, GithubProperty::class.java),

    /**
     * gitee
     */
    GITEE({ property -> gitee(property) }, GiteeProperty::class.java),

    /**
     * google
     */
    GOOGLE({ property -> google(property) }, GoogleProperty::class.java),

    /**
     * gitlab
     */
    GITLAB({ property -> gitlab(property) }, GitlabProperty::class.java),

    /**
     * weibo
     */
    WEIBO({ property -> weibo(property) }, WeiboProperty::class.java),

    /**
     * bccastle
     */
    BCCASTLE({ property -> bccastle(property) }, BccastleProperty::class.java),
    ;

    companion object {
        // ~=============================== Tools ===============================~ //
        /**
         * of
         *
         * @param name name
         * @return instance
         */
        fun of(name: String): Instance {
            for (instance in entries) {
                if (instance.toString() == name) {
                    return instance
                }
            }
            throw UiaException("not match instance for “$name”")
        }

        /**
         * of
         *
         * @param propertyClazz propertyClazz
         * @return instance
         */
        fun of(propertyClazz: Class<*>): Instance {
            for (instance in entries) {
                if (instance.propertyClazz == propertyClazz) {
                    return instance
                }
            }
            throw UiaException("un support property type : $propertyClazz")
        }
    }
}

// ~=============================== Instances ===============================~ //
/**
 * of yztoon
 *
 * @param property property
 * @return instance
 */
private fun yztoon(property: IUiaProperty): YztoonSdkInstance {
    return YztoonSdkInstance(property as YztoonProperty)
}

/**
 * of bjtoon
 *
 * @param property property
 * @return instance
 */
private fun bjtoon(property: IUiaProperty): ISdkInstance {
    return BjtoonSdkInstance(property as BjtoonProperty)
}

/**
 * of jban
 *
 * @param property property
 * @return instance
 */
private fun jban(property: IUiaProperty): JbanSdkInstance {
    return JbanSdkInstance(property as JbanProperty)
}

/**
 * of wx
 *
 * @param property property
 * @return instance
 */
private fun wx(property: IUiaProperty): WxSdkInstance {
    return WxSdkInstance(property as WxProperty)
}

/**
 * of alipay
 *
 * @param property property
 * @return instance
 */
private fun alipay(property: IUiaProperty): AlipaySdkInstance {
    return AlipaySdkInstance(property as AlipayProperty)
}

/**
 * of github
 *
 * @param property property
 * @return instance
 */
private fun github(property: IUiaProperty): GithubSdkInstance {
    return GithubSdkInstance(property as GithubProperty)
}

/**
 * of gitee
 *
 * @param property property
 * @return instance
 */
private fun gitee(property: IUiaProperty): GiteeSdkInstance {
    return GiteeSdkInstance(property as GiteeProperty)
}

/**
 * of google
 *
 * @param property property
 * @return instance
 */
private fun google(property: IUiaProperty): GoogleSdkInstance {
    return GoogleSdkInstance(property as GoogleProperty)
}

/**
 * of gitlab
 *
 * @param property property
 * @return instance
 */
private fun gitlab(property: IUiaProperty): GitlabSdkInstance {
    return GitlabSdkInstance(property as GitlabProperty)
}

/**
 * of weibo
 *
 * @param property property
 * @return instance
 */
private fun weibo(property: IUiaProperty): WeiboSdkInstance {
    return WeiboSdkInstance(property as WeiboProperty)
}

/**
 * of bccastle
 *
 * @param property property
 * @return instance
 */
private fun bccastle(property: IUiaProperty): BccastleSdkInstance {
    return BccastleSdkInstance(property as BccastleProperty)
}